airbyte-integrations/connectors/source-parcel-perform/source_parcel_perform/source.py

```diff
@@ -0,0 +1,235 @@
+            self.logger.info(response.json()["message"])
+
+    def get_updated_state(
+        self,
+        current_stream_state: MutableMapping[str, Any],
+        latest_record: Mapping[str, Any],
+    ) -> Mapping[str, any]:
+        return {
+            self.cursor_field: max(
+                latest_record.get(self.cursor_field),
+                current_stream_state.get(self.cursor_field, self.config["start_date"]),
+            )
+        }
+
+
+class ShipmentsDetails(Shipments, ABC):
+    shipment_ids_index = 0
+    state_checkpoint_interval = None
+    config = {}
+
+    def next_page_token(
+        self, response: requests.Response
+    ) -> Optional[Mapping[str, Any]]:
+        """
+        We pass each shipment_uuid as a next page token in order to make individual requests per shipment
+        :param response: the most recent response from the API
+        :return If there is another page in the result, a mapping (e.g: dict) containing information needed to query the next page in the response.
+                If there are no more pages in the result, return None.
+        """
+        self.shipment_ids_index += 1
+        if self.shipment_ids_index >= len(self.shipment_ids):
+            return {}
+        return self.shipment_ids_index
+
+    def request_params(
+        self,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> MutableMapping[str, Any]:
+        try:
+            return {"shipment_uuid": self.shipment_ids[self.shipment_ids_index]}
+        except IndexError:
+            return {}
+
+    def path(
+        self,
+        stream_state: Mapping[str, Any] = None,
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> str:
+        return "/v5/shipment/details/"
+
+    def parse_response(
+        self, response: requests.Response, **kwargs
+    ) -> Iterable[Mapping]:
+        """
+        :return an iterable containing each record in the response
+        """
+        try:
+            yield response.json()["data"]
+        except KeyError:
+            return None
+
+
+class SourceParcelPerform(AbstractSource):
+    def check_connection(self, logger, config) -> Tuple[bool, any]:
+        """
+        Check whether configuration is correct.
+
+        :param config:  the user-input config object conforming to the connector's spec.json
+        :param logger:  logger object
+        :return Tuple[bool, any]: (True, None) if the input config can be used to connect to the API successfully, (False, error) otherwise.
+        """
+        auth = ParcelPerformAuthenticator(
+            "https://api.parcelperform.com/auth/oauth/token/",
+            config["client_id"],
+            config["client_secret"],
+            "access_token",
+        )
+        if auth.get_access_token():
+            return True, None
```

Comment:
See other connectors `check_connection` function example. Most cases you need a try/except